File: backend/llm_service.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"time"
)

// LLMConfig LLM配置
type LLMConfig struct {
	Provider string // "claude", "openai", "gemini"
	APIKey   string
	Model    string
	BaseURL  string
}

// LLMService LLM服务
type LLMService struct {
	config LLMConfig
	client *http.Client
}

// Message 消息结构
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// NewLLMService 创建LLM服务
func NewLLMService() *LLMService {
	provider := os.Getenv("LLM_PROVIDER")
	if provider == "" {
		provider = "claude"
	}

	config := LLMConfig{
		Provider: provider,
		APIKey:   os.Getenv("LLM_API_KEY"),
		Model:    os.Getenv("LLM_MODEL"),
		BaseURL:  os.Getenv("LLM_BASE_URL"),
	}

	// 设置默认模型
	if config.Model == "" {
		switch provider {
		case "claude":
			config.Model = "claude-3-5-sonnet-20241022"
		case "openai":
			config.Model = "gpt-4-turbo"
		case "gemini":
			config.Model = "gemini-1.5-flash"
		}
	}

	return &LLMService{
		config: config,
		client: &http.Client{Timeout: 30 * time.Second},
	}
}

// Generate 生成计划
func (s *LLMService) Generate(systemPrompt, userPrompt string) (string, error) {
	if s.config.APIKey == "" {
		log.Println("⚠️  LLM_API_KEY 未设置，使用模拟数据")
		return s.generateMockPlan()
	}

	switch s.config.Provider {
	case "claude":
		return s.callClaude(systemPrompt, userPrompt)
	case "openai":
		return s.callOpenAI(systemPrompt, userPrompt)
	case "gemini":
		return s.callGemini(systemPrompt, userPrompt)
	default:
		return "", fmt.Errorf("unsupported provider: %s", s.config.Provider)
	}
}

// callClaude 调用Claude API
func (s *LLMService) callClaude(systemPrompt, userPrompt string) (string, error) {
	url := "https://api.anthropic.com/v1/messages"

	payload := map[string]interface{}{
		"model":       s.config.Model,
		"max_tokens":  2000,
		"system":      systemPrompt,
		"messages": []map[string]string{
			{
				"role":    "user",
				"content": userPrompt,
			},
		},
	}

	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(body))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", s.config.APIKey)
	req.Header.Set("anthropic-version", "2023-06-01")

	resp, err := s.client.Do(req)
	if err != nil {
		log.Printf("❌ Claude API 调用失败: %v", err)
		return "", err
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)

	if resp.StatusCode != http.StatusOK {
		log.Printf("❌ Claude API 返回错误: %s", string(respBody))
		return "", fmt.Errorf("claude api error: %s", string(respBody))
	}

	var result map[string]interface{}
	json.Unmarshal(respBody, &result)

	// 提取响应内容
	if content, ok := result["content"].([]interface{}); ok && len(content) > 0 {
		if msg, ok := content[0].(map[string]interface{}); ok {
			if text, ok := msg["text"].(string); ok {
				return text, nil
			}
		}
	}

	return "", fmt.Errorf("invalid claude response format")
}

// callOpenAI 调用OpenAI API
func (s *LLMService) callOpenAI(systemPrompt, userPrompt string) (string, error) {
	url := "https://api.openai.com/v1/chat/completions"

	payload := map[string]interface{}{
		"model":       s.config.Model,
		"temperature": 0.7,
		"messages": []map[string]string{
			{"role": "system", "content": systemPrompt},
			{"role": "user", "content": userPrompt},
		},
	}

	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(body))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+s.config.APIKey)

	resp, err := s.client.Do(req)
	if err != nil {
		log.Printf("❌ OpenAI API 调用失败: %v", err)
		return "", err
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)

	if resp.StatusCode != http.StatusOK {
		log.Printf("❌ OpenAI API 返回错误: %s", string(respBody))
		return "", fmt.Errorf("openai api error: %s", string(respBody))
	}

	var result map[string]interface{}
	json.Unmarshal(respBody, &result)

	// 提取响应内容
	if choices, ok := result["choices"].([]interface{}); ok && len(choices) > 0 {
		if choice, ok := choices[0].(map[string]interface{}); ok {
			if msg, ok := choice["message"].(map[string]interface{}); ok {
				if text, ok := msg["content"].(string); ok {
					return text, nil
				}
			}
		}
	}

	return "", fmt.Errorf("invalid openai response format")
}

// callGemini 调用Gemini API
func (s *LLMService) callGemini(systemPrompt, userPrompt string) (string, error) {
	url := fmt.Sprintf("https://generativelanguage.googleapis.com/v1beta/models/%s:generateContent?key=%s", s.config.Model, s.config.APIKey)

	payload := map[string]interface{}{
		"system_instruction": map[string]interface{}{
			"parts": []map[string]string{
				{"text": systemPrompt},
			},
		},
		"contents": []map[string]interface{}{
			{
				"parts": []map[string]string{
					{"text": userPrompt},
				},
			},
		},
	}

	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(body))
	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(req)
	if err != nil {
		log.Printf("❌ Gemini API 调用失败: %v", err)
		return "", err
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)

	if resp.StatusCode != http.StatusOK {
		log.Printf("❌ Gemini API 返回错误: %s", string(respBody))
		return "", fmt.Errorf("gemini api error: %s", string(respBody))
	}

	var result map[string]interface{}
	json.Unmarshal(respBody, &result)

	// 提取响应内容
	if candidates, ok := result["candidates"].([]interface{}); ok && len(candidates) > 0 {
		if candidate, ok := candidates[0].(map[string]interface{}); ok {
			if content, ok := candidate["content"].(map[string]interface{}); ok {
				if parts, ok := content["parts"].([]interface{}); ok && len(parts) > 0 {
					if part, ok := parts[0].(map[string]interface{}); ok {
						if text, ok := part["text"].(string); ok {
							return text, nil
						}
					}
				}
			}
		}
	}

	return "", fmt.Errorf("invalid gemini response format")
}

// generateMockPlan 生成模拟计划（用于测试）
func (s *LLMService) generateMockPlan() (string, error) {
	mockPlan := `{
  "date": "2025-10-27",
  "summary": "今日学习计划已生成",
  "tasks": [
    {
      "goal_id": 1,
      "title": "Go语言基础学习",
      "description": "学习Go语言的并发编程和goroutine",
      "duration_minutes": 90,
      "start_time": "09:00",
      "priority": "high"
    },
    {
      "goal_id": 2,
      "title": "React组件开发",
      "description": "完成TDesign组件集成和优化",
      "duration_minutes": 60,
      "start_time": "14:00",
      "priority": "medium"
    },
    {
      "goal_id": 3,
      "title": "数据库设计复习",
      "description": "复习TDSQL-C MySQL的最佳实践",
      "duration_minutes": 45,
      "start_time": "16:00",
      "priority": "medium"
    }
  ],
  "total_duration_minutes": 195,
  "notes": "这是一个模拟计划，用于测试。请设置LLM_API_KEY环境变量以使用真实的LLM服务。"
}`
	return mockPlan, nil
}

File: backend-py/app/services/llm_service.py
import os
from langchain.prompts import ChatPromptTemplate
from langchain.schema.output_parser import StrOutputParser
from langchain_openai import ChatOpenAI

class LLMService:
    def __init__(self):
        """初始化LLM服务，按优先级选择可用的LLM提供商"""
        self.llm = self._init_llm()
        self.output_parser = StrOutputParser()
    
    def _init_llm(self):
        """初始化LLM模型，按优先级尝试"""
        # 首先尝试Claude
        claude_api_key = os.getenv("ANTHROPIC_API_KEY")
        if claude_api_key:
            try:
                return ChatAnthropic(model="claude-3-opus-20240229", api_key=claude_api_key)
            except Exception as e:
                print(f"Claude初始化失败: {e}")
        
        # 然后尝试OpenAI
        openai_api_key = os.getenv("OPENAI_API_KEY")
        if openai_api_key:
            try:
                return ChatOpenAI(model="gpt-4-turbo", api_key=openai_api_key)
            except Exception as e:
                print(f"OpenAI初始化失败: {e}")
        
        # 最后尝试Gemini
        gemini_api_key = os.getenv("GOOGLE_API_KEY")
        if gemini_api_key:
            try:
                return ChatGoogleGenerativeAI(model="gemini-pro", api_key=gemini_api_key)
            except Exception as e:
                print(f"Gemini初始化失败: {e}")
        
        # 如果没有配置任何API Key，返回None
        print("warning: 未配置任何LLM API Key，将使用模拟数据")
        return None
    
    def generate(self, prompt: str) -> str:
        """生成LLM响应"""
        if not self.llm:
            # 返回模拟数据
            return self._generate_mock_response(prompt)
        
        try:
            # 使用LangChain的链来处理
            chat_prompt = ChatPromptTemplate.from_template("{prompt}")
            chain = chat_prompt | self.llm | self.output_parser
            response = chain.invoke({"prompt": prompt})
            return response
        except Exception as e:
            print(f"LLM生成失败: {e}")
            # 失败时返回模拟数据
            return self._generate_mock_response(prompt)
    
    def _generate_mock_response(self, prompt: str) -> str:
        """生成模拟的LLM响应"""
        return """
{
  "date": "2025-10-27",
  "summary": "今日学习计划已生成",
  "tasks": [
    {
      "goal_id": 1,
      "title": "Go语言基础学习",
      "description": "学习Go语言的并发编程和goroutine",
      "duration_minutes": 90,
      "start_time": "09:00",
      "priority": "high"
    },
    {
      "goal_id": 2,
      "title": "React组件开发",
      "description": "完成TDesign组件集成和优化",
      "duration_minutes": 60,
      "start_time": "14:00",
      "priority": "medium"
    },
    {
      "goal_id": 3,
      "title": "数据库设计复习",
      "description": "复习TDSQL-C MySQL的最佳实践",
      "duration_minutes": 45,
      "start_time": "16:00",
      "priority": "medium"
    }
  ],
  "total_duration_minutes": 195,
  "notes": "这是一个模拟计划，用于测试。请设置LLM_API_KEY环境变量以使用真实的LLM服务。"
}
"""